Lib/ipaddress.py

```diff
@@ -1817,6 +1835,8 @@ def __init__(self, address):
             AddressValueError: If address isn't a valid IPv6 address.
 
         """
+        self._scope_id = None
```

Comment:
What's the reason for setting this up here? We just set it again below when we actually get the data, right?

... Ah I see, this sets it for the `int` and `bytes` cases below.

I think it'd be better to be explicit: write `self._scope_id = None` (well, actually `self._zone_id = None`) in each of those two cases. That makes it super clear that in those cases the zone ID is always None. And it only costs one extra line of code! :wink:

Reply:
As far as I know, it is the best practice to avoid attribute initialization in `if `statement. According to this and [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), I prefer to leave it as is.
If it creates any issues, I will definitely fix it.

Reply:
I think following DRY here is the right thing to do, and it's not at all uncommon to initialize default values for attributes in ``__init__``. Generally, DRY isn't heavily prioritized when it's just two repetitions. But if more cases/conditional blocks are added in the future, writing ``self._scope_id = None`` (or ``self._zone_id``) in all of them could get quite repetitive.

Reply:
Hmm, I find my feeling is I would actually much rather write this explicitly in the two `if` cases.

Some thoughts that perhaps explain why:

* This isn't a default value. It's the value that applies in two possible cases... specifically the two cases that are sort of off to the side and special. In the case that (a) is conceptually the main one, as evidenced for example by appearing in virtually all of the examples in the module's docs, and (b) contains the most code, the value set here always gets clobbered by a later assignment.

  So when you're reading this function and just want to first understand the main, normal, case, this line is a red herring because it never has an effect.

  (It happens that in the main case, this value also often/usually ends up being `None`. But that just makes having this line more confusing, because it *looks* so much like it's what causes that to happen, when in fact that's entirely controlled by `_split_scope_id`.)

* It's true that there are two cases where this value applies. But that's more like two facts than one fact: the `int` form of the constructor doesn't take a scope zone ID, and the `bytes` form doesn't take one, though the main `str` (or str-able) form of the constructor does.

  One way of seeing that: if we imagine that some fourth form were to be added to this constructor in the future, it's not at all obvious whether that form would accept a scope zone ID or wouldn't.

  Because this is two independent facts about two different forms of the constructor, it's best to express each of those facts separately.

  If the code expressing this fact were more complicated, I'd certainly be in favor of writing down that complexity once... and then I'd again want two separate lines to invoke it (e.g. to call some helper method) for the two different cases.

Reply:
> This isn't a default value. It's the value that applies in two possible cases... specifically the two cases that are sort of off to the side and special.

I may have been a bit presumptuous then and assumed that ``self._scope_id`` being ``None`` was a significantly more common scenario. If it is fairly uncommon that the address passed into the ``IPv6Address`` class will be an integer or byte, and it is unlikely that other conditional blocks would set it to ``None``, I would agree with your argument to explicitly declare it within those two blocks instead of its current position.

So just to clarify this suggestion, would it look something like this? I omitted the comments to make it more succinct.

```python3
def __init__(self, address):
        if isinstance(address, int):
            self._scope_id = None
            self._check_int_address(address)
            self._ip = address
            return

        if isinstance(address, bytes):
            self._scope_id = None
            self._check_packed_address(address, 16)
            self._ip = int.from_bytes(address, 'big')
            return

```

Reply:
> So just to clarify this suggestion, would it look something like this?

Yep -- just like that except that I'd probably put the `_scope_id` line just after the `_ip` line, rather than before:

        if isinstance(address, int):
            self._check_int_address(address)
            self._ip = address
            self._scope_id = None
            return

(and similarly in the other block.)

I'd lean that way because:
* the `_ip` part is the main thing this code is about, and the `_scope_id` is the extra added feature that's sometimes important; so leading with the main thing makes it quicker to follow the main logic and tune out the rest
* in the `...%...` syntax where they both appear (the one handled in the main codepath, below), the part corresponding to `_scope_id` comes after the part corresponding to `_ip`. So that makes it natural to think of `_ip` as coming first, then `_scope_id`.

Reply:
Fixed.